sdk/cosmosdb/cosmos/src/retry/timeoutFailoverRetryPolicy.ts

```diff
@@ -0,0 +1,125 @@
+import { RetryContext } from "./RetryContext";
+import { CosmosHeaders } from "../queryExecutionContext/CosmosHeaders";
+import { TimeoutErrorCode } from "../request/TimeoutError";
+import { ErrorResponse } from "../request";
+
+/**
+ * This class TimeoutFailoverRetryPolicy handles retries for read operations
+ * (including data plane,metadata, and query plan) in case of request timeouts
+ * (TimeoutError) or service unavailability (503 status code) by performing failover
+ * and retrying on other regions.
+ * @hidden
+ */
+export class TimeoutFailoverRetryPolicy implements RetryPolicy {
+  private maxRetryAttemptCount = 120;
+  private maxServiceUnavailableRetryCount = 1;
+  public retryAfterInMs = 0;
+  public failoverRetryCount = 0;
+  public request: any;
+  public locationEndpoint: any;
+
+  constructor(
+    private globalEndpointManager: GlobalEndpointManager,
+    private headers: CosmosHeaders,
+    private methodType: HTTPMethod,
+    private resourceType: ResourceType,
+    private operationType: OperationType,
+    private enableEndPointDiscovery: boolean
+  ) {}
+
+  /**
+   * Checks if a timeout request is valid for the timeout failover retry policy.
+   * A valid request should be a data plane, metadata, or query plan request.
+   * @returns
+   */
+  private isValidRequestForTimeoutError(): boolean {
+    const isQuery = Constants.HttpHeaders.IsQuery in this.headers;
+    const isQueryPlan = Constants.HttpHeaders.IsQueryPlan in this.headers;
+    if (this.methodType === HTTPMethod.get || isQuery || isQueryPlan) {
+      return true;
+    }
+    return false;
+  }
+
+  public async shouldRetry(
+    err: ErrorResponse,
+    retryContext?: RetryContext,
+    locationEndpoint?: string
+  ): Promise<boolean> {
+    if (!err) {
+      return false;
+    }
+    if (!retryContext || !locationEndpoint) {
+      return false;
+    }
+    // Check if the error is a timeout error (TimeoutErrorCode) and if it is not a valid HTTP network timeout request
+    if (err.code === TimeoutErrorCode && !this.isValidRequestForTimeoutError()) {
+      return false;
+    }
+    if (!this.enableEndPointDiscovery) {
+      return false;
+    }
+    if (
+      err.code === StatusCodes.ServiceUnavailable &&
+      this.failoverRetryCount >= this.maxServiceUnavailableRetryCount
+    ) {
+      return false;
+    }
+    if (this.failoverRetryCount >= this.maxRetryAttemptCount) {
+      return false;
+    }
+    const canUseMultipleWriteLocations = this.globalEndpointManager.canUseMultipleWriteLocations(
+      this.resourceType,
+      this.operationType
+    );
+    const readRequest = isReadRequest(this.operationType);
+
+    if (!canUseMultipleWriteLocations && !readRequest) {
+      // Write requests on single master cannot be retried, no other regions available
+      return false;
+    }
+    this.failoverRetryCount++;
+    // Setting the retryLocationIndex to the next available location for retry.
```

Comment:
Why are not marking the location unavailable and using the index to move to next location here? As we have done in other policies i.e in `endpointDiscoveryRetryPolicy.ts`

Reply:
Timeouts are not signs that the region is unavailable, this aligns with other Cosmos SDKs. There are specific response codes that signal a region becoming/is unavailable, timeouts are not one of them anywhere

Reply:
The goal based on other SDKs and the design document linked is to align the JS SDK with other SDKs. Other SDKs whenever there are timeouts, attempt to retry, that single request, in another region, under the assumption that the transient network condition that generated the timeout might not affect another network path. It's a best effort to provide availability.

Reply:
I understand the reasoning behind the implementation of timeouts. However, @ealsur, could you please assist us in understanding the rationale for not marking the region as unavailable for 503 errors? Is it just alignment with other SDK's?

Reply:
Because a 503 is not a signal that the region is unavailable. The service contract does not include 503.

503's meaning across our public documentation (https://docs.microsoft.com/en-us/azure/cosmos-db/sql/conceptual-resilient-sdk-applications#timeouts-and-connectivity-related-failures-http-408503, https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-sdk-availability#transient-connectivity-issues-on-tcp-protocol) is that it's a transient problem.

Reply:
Got it! I appreciate the clarification, @ealsur. This information is incredibly useful.